libs/small-steps/src/Control/Iterate/BiMap.hs

```diff
@@ -0,0 +1,144 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeApplications #-}
+
+module Control.Iterate.BiMap where
+
+import Cardano.Binary
+  ( Decoder,
+    FromCBOR (..),
+    ToCBOR (..),
+    decodeListLen,
+    decodeMapSkel,
+    dropMap,
+  )
+import Codec.CBOR.Encoding (encodeListLen)
+import Control.DeepSeq (NFData (rnf))
+-- import Data.List (sortBy)
+-- import qualified Data.List as List
+-- import Data.Set (Set)
+
+import Control.Monad (void)
+import Data.Coders (invalidKey)
+import Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import qualified Data.Set as Set
+import NoThunks.Class (NoThunks (..))
+
+-- =================== Basic BiMap =====================
+-- For Bijections we define (BiMap v k v).  Reasons we can't use (Data.Bimap k v)
+-- 1) We need to enforce that the second argument `v` is in the Ord class, when making it an Iter instance.
+-- 2) The constructor for Data.BiMap is not exported, and it implements a Bijection
+-- 3) Missing operation 'restrictkeys' and 'withoutkeys' make performant versions of operations  ◁ ⋪ ▷ ⋫ hard.
+-- 4) Missing operation 'union', make performant versions of ∪ and ⨃ hard.
+-- 5) So we roll our own which is really a (Data.Map k v) with an index that maps v to Set{k}
+
+data BiMap v a b where MkBiMap :: (v ~ b) => !(Map.Map a b) -> !(Map.Map b (Set.Set a)) -> BiMap v a b
+
+--  ^   the 1st and 3rd parameter must be the same:             ^   ^
+
+biMapToMap :: BiMap v a b -> Map a b
+biMapToMap (MkBiMap m _) = m
+
+-- ============== begin necessary Cardano.Binary instances ===============
+instance (Ord a, Ord b, ToCBOR a, ToCBOR b) => ToCBOR (BiMap b a b) where
+  -- The `toCBOR` instance encodes only the forward map. We wrap this in a
+  -- length-one list because a _previous_ encoding wrote out both maps, and we
+  -- can easily use the list length token to distinguish between them.
+  toCBOR (MkBiMap l _) = encodeListLen 1 <> toCBOR l
+
+instance
+  forall a b.
+  (Ord a, Ord b, FromCBOR a, FromCBOR b) =>
+  FromCBOR (BiMap b a b)
+  where
+  fromCBOR =
+    decodeListLen >>= \case
+      1 -> decodeMapAsBimap
+      -- Previous encoding of 'BiMap' encoded both the forward and reverse
+      -- directions. In this case we skip the reverse encoding. Note that,
+      -- further, the reverse encoding was from 'b' to 'a', not the current 'b'
+      -- to 'Set a', and hence the dropper reflects that.
+      2 -> do
+        !x <- decodeMapAsBimap
+        dropMap (void $ fromCBOR @b) (void $ fromCBOR @a)
+        return x
+      k -> invalidKey (fromIntegral k)
+
+-- | Decode a serialised CBOR Map as a Bimap
+decodeMapAsBimap ::
+  (FromCBOR a, FromCBOR b, Ord a, Ord b) =>
+  Decoder s (BiMap b a b)
+decodeMapAsBimap = decodeMapSkel biMapFromAscDistinctList
```

Comment:
This function is broken again, see the fix in 7ef96a683bc0ce6ab1be25ac0ced5c212dea90a7

Reply:
Good catch. I didn't realise this. I have fixed it back.